lib/openstack_query/query_blocks/query_grouper.py

```diff
@@ -0,0 +1,137 @@
+from typing import List, Dict, Optional, Union
+import logging
+from collections import OrderedDict
+from openstack_query.query_blocks.result import Result
+from enums.query.props.prop_enum import PropEnum
+from custom_types.openstack_query.aliases import GroupMappings, GroupRanges, PropValue
+from exceptions.parse_query_error import ParseQueryError
+
+logger = logging.getLogger(__name__)
+
+
+class QueryGrouper:
+    """
+    Helper class for implementing grouping on query outputs
+    """
+
+    def __init__(self, prop_enum_cls):
+        self._prop_enum_cls = prop_enum_cls
+        self._group_by = None
+        self._group_mappings = {}
+
+    def _build_unique_val_groups(
+        self,
+        obj_list: List,
+    ) -> GroupMappings:
+        """
+        helper method to find all unique values for a given property in query results, and then,
+        for each unique value, create a group mapping
+        :param obj_list: A list of openstack objects to build unique groups for
+        """
+        prop_func = self._prop_enum_cls.get_prop_mapping(self._group_by)
+        # ordered dict to mimic ordered set
+        # this is to preserve order we see unique values in - in case a sort has been done already
+        unique_vals = OrderedDict({prop_func(obj): None for obj in obj_list})
+        logger.debug(
+            "unique values found %s - each will become a group",
+            ",".join(f"{val}" for val in unique_vals),
+        )
+        group_mappings = {}
+        # build groups
+        for val in unique_vals.keys():
+            group_key = val
+            group_mappings[group_key] = (
+                lambda obj, test_val=val: prop_func(obj) == test_val
+            )
+        return group_mappings
+
+    def _add_include_missing_group(self, group_ranges: GroupRanges):
+        """
+        Helper method used to add an extra group for group_by which includes
+        all resource values not already included in any defined mappings
+        :param group_ranges: A set of predefined group ranges to check against
+        """
+        # if ungrouped group wanted - filter for all not in any range specified
+        all_prop_list = set()
+        for _, prop_list in group_ranges.items():
+            all_prop_list.update(set(prop_list))
```

Comment:
Can we use .values() here instead of .items(). I see you are using `_` as an ignored parameter.
```suggestion
        for prop_list in group_ranges.values():
            all_prop_list.update(set(prop_list))
```